### PR TITLE
block_with_senders in ethstatecache

### PR DESCRIPTION
**Pull request by supernovahs**
Ref #5090
TODO:
Change conversion function to use the api 

@mattsse Can u take a look at the changes and suggest a better approach?
My code is doing a lot of clones.

### CODE REVIEW

File: crates/rpc/rpc-types/src/eth/block.rs [hunk 1]
```diff
@@ -486,25 +486,25 @@ impl<'de> Deserialize<'de> for BlockId {
                     match key.as_str() {
                         "blockNumber" => {
                             if number.is_some() || block_hash.is_some() {
-                                return Err(serde::de::Error::duplicate_field("blockNumber"));
+                                return Err(serde::de::Error::duplicate_field("blockNumber"))
```
Comment (mattsse) [inline crates/rpc/rpc-types/src/eth/block.rs:489]: please don't reformat this, this makes it a lot harder to review

File: crates/rpc/rpc/src/eth/cache/mod.rs [hunk 2]
```diff
@@ -331,11 +369,54 @@ where
...
+                                                let _ = action_tx.send(
+                                                    CacheAction::BlockWithSendersResult {
+                                                        block_hash,
+                                                        res: block_sender,
+                                                    },
+                                                );
+                                            })
+                                            .unwrap_or_else(|| {})
+                                    })
+                                    .unwrap_or_else(|e| {});
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/cache/mod.rs:416]: these unwraps seem unneccessary?

File: crates/rpc/rpc/src/eth/cache/mod.rs [hunk 3]
```diff
@@ -331,11 +369,54 @@ where
...
+                                let provider = this.provider.clone();
+                                let action_tx = this.action_tx.clone();
+                                let rate_limiter = this.rate_limiter.clone();
+                                this.action_task_spawner.spawn_blocking(Box::pin(async move {
+                                    // Acquire permit
+                                    let _permit = rate_limiter.acquire().await;
+                                    // Only look in the database to prevent situations where we
+                                    // looking up the tree is blocking
+                                    let res = provider
+                                        .find_block_by_hash(block_hash, BlockSource::Database);
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/cache/mod.rs:398]: this step is redundant we only do this to get the block number.

we should add a new helper function to the provider that takes the hash instead:

`block_with_senders_by_hash`:

https://github.com/paradigmxyz/reth/blob/main/crates/storage/provider/src/traits/block.rs#L115-L124

File: crates/rpc/rpc/src/eth/cache/mod.rs [hunk 4]
```diff
@@ -29,24 +31,35 @@ mod metrics;
...
+    Block(BlockResponseSender),
+    BlockTransactions(BlockTransactionsResponseSender),
+    BlockWithSenders(BlockWithSendersResponseSender),
+}
+
+#[derive(Debug)]
+enum BlockData {
+    Block(Block),
+    BlockWithSenders(BlockWithSenders),
+}
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/cache/mod.rs:45]: I think we can simplify this by a lot if we always store `BlockWithSenders` instead, so every missing block should be fetched with senders

**Review by mattsse — CHANGES_REQUESTED** [review]
this looks great already!

some requests:
* modify `block_with_senders` so that it supports `BlockHashOrNumber` (should be separate PR)
* always store BlockWithSenders

**Comment by supernovahs** [comment]
https://github.com/paradigmxyz/reth/blob/63f67d97fd7e0007c806efb58dfae04d4aa8d895/crates/rpc/rpc-types-compat/src/block.rs#L68-L77 
This is the bottleneck right?

**Comment by mattsse** [comment]
@supernovahs do you have the bandwidth to get this over the line?

**Comment by supernovahs** [comment]
At the stage of changing the conversion function. Trying to figure out how to use the senders in from_block_full

**Comment by supernovahs** [comment]
Let me know if this implementation makes sense. Thanks
@mattsse

File: crates/primitives/src/transaction/mod.rs [hunk 5]
```diff
@@ -1397,11 +1397,11 @@ impl<'a> arbitrary::Arbitrary<'a> for TransactionSigned {
 #[derive(Debug, Clone, PartialEq, Hash, Eq, AsRef, Deref, Default)]
 pub struct TransactionSignedEcRecovered {
     /// Signer of the transaction
-    signer: Address,
+    pub signer: Address,
     /// Signed transaction
     #[deref]
     #[as_ref]
-    signed_transaction: TransactionSigned,
+    pub signed_transaction: TransactionSigned,
```
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:1404]: these should remain private

File: crates/rpc/rpc-types-compat/src/block.rs [hunk 6]
```diff
@@ -51,35 +54,38 @@ pub fn from_block_with_tx_hashes(
...
-            signed_tx,
-            block_hash,
-            block_number,
-            base_fee_per_gas,
-            U256::from(idx),
-        ))
-    }
+    let block_length = block.block.length();
+    let body = std::mem::take(&mut block.block.body);
+    let transactions_with_senders = body.into_iter().zip(block.senders);
```
Comment (mattsse) [inline crates/rpc/rpc-types-compat/src/block.rs:69]: there are now helper functions for this on the block type

File: crates/rpc/rpc-types-compat/src/block.rs [hunk 7]
```diff
@@ -51,35 +54,38 @@ pub fn from_block_with_tx_hashes(
...
+                TransactionSignedEcRecovered { signer: sender, signed_transaction: tx };
+            Ok(from_recovered_with_block_context(
+                signed_tx_ec_recovered,
+                block_hash,
+                block_number,
+                base_fee_per_gas,
+                U256::from(idx),
+            ))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
```
Comment (mattsse) [inline crates/rpc/rpc-types-compat/src/block.rs:83]: we don't need the result anymore, right?

File: crates/rpc/rpc-types-compat/src/block.rs [hunk 8]
```diff
@@ -51,35 +54,38 @@ pub fn from_block_with_tx_hashes(
...
-        ))
-    }
+    let block_length = block.block.length();
+    let body = std::mem::take(&mut block.block.body);
+    let transactions_with_senders = body.into_iter().zip(block.senders);
+    let transactions = transactions_with_senders
+        .enumerate()
+        .map(|(idx, (tx, sender))| {
+            let signed_tx_ec_recovered =
+                TransactionSignedEcRecovered { signer: sender, signed_transaction: tx };
```
Comment (mattsse) [inline crates/rpc/rpc-types-compat/src/block.rs:74]: there's now, tx.with_signer()

File: crates/rpc/rpc/src/eth/cache/mod.rs [hunk 9]
```diff
@@ -29,12 +31,18 @@ mod metrics;
 mod multi_consumer;
 pub use multi_consumer::MultiConsumerLruCache;
 
-/// The type that can send the response to a requested [Block]
-type BlockResponseSender = oneshot::Sender<RethResult<Option<Block>>>;
+#[derive(Debug)]
+enum BlockData {
+    Block(Block),
+    BlockWithSenders(BlockWithSenders),
+}
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/cache/mod.rs:38]: we need to get rid of this enum and only handle BlockWithSenders

File: crates/rpc/rpc/src/eth/cache/mod.rs [hunk 10]
```diff
@@ -431,7 +480,10 @@ where
                         }
                         CacheAction::CacheNewCanonicalChain { blocks, receipts } => {
                             for block in blocks {
-                                this.on_new_block(block.hash, Ok(Some(block.unseal())));
+                                this.on_new_block(
+                                    block.hash,
+                                    Ok(Some(BlockData::Block(block.unseal()))),
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/cache/mod.rs:485]: this needs to insert the BlockWithsenders,

we need to update:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/cache/mod.rs#L534-L534

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, this is almost there.

File: crates/rpc/rpc/src/eth/api/block.rs [hunk 11]
```diff
@@ -184,7 +184,12 @@ where
             .provider()
             .header_td_by_number(block.number)?
             .ok_or(EthApiError::UnknownBlockNumber)?;
-        let block = from_block(block.into(), total_difficulty, full.into(), Some(block_hash))?;
-        Ok(Some(block.into()))
+        let block_with_senders = self.cache().get_block_with_senders(block_hash).await?;
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/block.rs:187]: this is redundant, because we already fetch the block above?

**Review by mattsse — CHANGES_REQUESTED** [review]
cache changes lgtm

still a few nits + failing tests

**Comment by mattsse** [comment]
needs rebase and two open comments, then this is good

**Comment by mattsse** [comment]
bump @supernovahs 
I really want to get this in

**Comment by supernovahs** [comment]
Ah darn , I am afk for atleast 6 hours.
I tried to figure to out the bug as to why tests were failing. But no progress on it yet

**Review by mattsse — APPROVED** [review]
great!

I added the missing parts

**Comment by supernovahs** [comment]
thanks 🫡

**Pull request merged**